onmt/transforms/terminology.py

```diff
@@ -0,0 +1,346 @@
+            "-term_example_ratio",
+            type=float,
+            default=0.2,
+            help="Max terms allowed in an example.",
+        )
+        group.add(
+            "--src_term_stoken",
+            "-src_term_stoken",
+            type=str,
+            help="The source term start token.",
+            default="｟src_term_start｠",
+        )
+        group.add(
+            "--tgt_term_stoken",
+            "-tgt_term_stoken",
+            type=str,
+            help="The target term start token.",
+            default="｟tgt_term_start｠",
+        )
+        group.add(
+            "--tgt_term_etoken",
+            "-tgt_term_etoken",
+            type=str,
+            help="The target term end token.",
+            default="｟tgt_term_end｠",
+        )
+        group.add(
+            "--term_source_delimiter",
+            "-term_source_delimiter",
+            type=str,
+            help="Any special token used for augmented source sentences. "
+            "The default is the fuzzy token used in the "
+            "FuzzyMatch transform.",
+            default="｟fuzzy｠",
+        )
+
+    def _parse_opts(self):
+        self.termbase_path = self.opts.termbase_path
+        self.src_spacy_language_model = self.opts.src_spacy_language_model
+        self.tgt_spacy_language_model = self.opts.tgt_spacy_language_model
+        self.term_corpus_ratio = self.opts.term_corpus_ratio
+        self.term_example_ratio = self.opts.term_example_ratio
+        self.term_source_delimiter = self.opts.term_source_delimiter
+        self.src_term_stoken = self.opts.src_term_stoken
+        self.tgt_term_stoken = self.opts.tgt_term_stoken
+        self.tgt_term_etoken = self.opts.tgt_term_etoken
+
+    @classmethod
+    def get_specials(cls, opts):
+        """Add the term tokens to the src vocab."""
+        src_specials = list()
+        src_specials.extend(
+            [opts.src_term_stoken, opts.tgt_term_stoken, opts.tgt_term_etoken]
+        )
+        return (src_specials, list())
+
+    def warm_up(self, vocabs=None):
+        """Create the terminology matcher."""
+
+        super().warm_up(None)
+        self.termmatcher = TermMatcher(
+            self.termbase_path,
+            self.src_spacy_language_model,
+            self.tgt_spacy_language_model,
+            self.term_example_ratio,
+            self.src_term_stoken,
+            self.tgt_term_stoken,
+            self.tgt_term_etoken,
+            self.term_source_delimiter,
+            self.term_corpus_ratio,
+        )
+
+    def batch_apply(self, batch, is_train=False, stats=None, **kwargs):
+        bucket_size = len(batch)
+        examples_with_terms = 0
+
+        for i, (ex, _, _) in enumerate(batch):
+            # Skip half examples to improve performance. This means we set
+            # a hard limit for the `term_corpus_ratio` to 0.5, which is actually
+            # quite high. TODO: We can add this (skipping examples) as an option
```

Comment:
What kind of perf issues are we talking about if we apply this fully? (The current implementation is not super "random" so depending on how the data is looped on it might not be fully optimized augmentation.)

Reply:
Also, if absolutely necessary, maybe make this configurable.

Reply:
> What kind of perf issues are we talking about if we apply this fully? (The current implementation is not super "random" so depending on how the data is looped on it might not be fully optimized augmentation.)

The transform adds a large overhead when every single example is passed to the term matching function and the whole training requires many workers (and lots of RAM) to keep feeding the GPU with samples. It's still under investigation but for the time being it seems to alleviate the problem.
Could you clarify a bit the part about the "randomness"?

Reply:
If we don't skip one out of two example, it makes things very slow when rebuilding a bucket so this trick is just for performance but indeed imply a hard limit of 50% max of examples to be tagged. It is not necessary to make this variable since the actual ratio is already an option. I think it is good as is.